File: src/Features/VisualBasic/Portable/Completion/CompletionProviders/OverrideCompletionProvider.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.Collections.Immutable
Imports System.Composition
Imports System.Threading
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Completion
Imports Microsoft.CodeAnalysis.Completion.Providers
Imports Microsoft.CodeAnalysis.Editing
Imports Microsoft.CodeAnalysis.Host.Mef
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.Providers
    <ExportCompletionProvider(NameOf(OverrideCompletionProvider), LanguageNames.VisualBasic)>
    <ExtensionOrder(After:=NameOf(CompletionListTagCompletionProvider))>
    <[Shared]>
    Friend Class OverrideCompletionProvider
        Inherits AbstractOverrideCompletionProvider

        Private _isFunction As Boolean
        Private _isSub As Boolean
        Private _isProperty As Boolean

        <ImportingConstructor>
        <Obsolete(MefConstruction.ImportingConstructorMessage, True)>
        Public Sub New()
        End Sub

        Friend Overrides ReadOnly Property Language As String
            Get
                Return LanguageNames.VisualBasic
            End Get
        End Property

        Protected Overrides Function GetSyntax(commonSyntaxToken As SyntaxToken) As SyntaxNode
            Dim token = CType(commonSyntaxToken, SyntaxToken)

            Dim propertyBlock = token.GetAncestor(Of PropertyBlockSyntax)()
            If propertyBlock IsNot Nothing Then
                Return propertyBlock
            End If

            Dim methodBlock = token.GetAncestor(Of MethodBlockBaseSyntax)()
            If methodBlock IsNot Nothing Then
                Return methodBlock
            End If

            Return token.GetAncestor(Of MethodStatementSyntax)()
        End Function

        Protected Overrides Function GetToken(completionItem As CompletionItem, syntaxTree As SyntaxTree, cancellationToken As CancellationToken) As SyntaxToken
            Dim tokenSpanEnd = MemberInsertionCompletionItem.GetTokenSpanEnd(completionItem)
            Return syntaxTree.FindTokenOnLeftOfPosition(tokenSpanEnd, cancellationToken)
        End Function

        Public Overrides Function FindStartingToken(syntaxTree As SyntaxTree, position As Integer, cancellationToken As CancellationToken) As SyntaxToken
            Dim token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken)
            Return token.GetPreviousTokenIfTouchingWord(position)
        End Function

        Public Overrides Function IsInsertionTrigger(text As SourceText, characterPosition As Integer, options As CompletionOptions) As Boolean
            Return CompletionUtilities.IsTriggerAfterSpaceOrStartOfWordCharacter(text, characterPosition, options)
        End Function

        Public Overrides ReadOnly Property TriggerCharacters As ImmutableHashSet(Of Char) = CompletionUtilities.SpaceTriggerChar

        Public Overrides Function TryDetermineModifiers(startToken As SyntaxToken,
                                                        text As SourceText, startLine As Integer,
                                                        ByRef seenAccessibility As Accessibility,
                                                        ByRef modifiers As DeclarationModifiers) As Boolean

            Dim token = CType(startToken, SyntaxToken)
            modifiers = New DeclarationModifiers()
            seenAccessibility = Accessibility.NotApplicable
            Dim overridesToken = New SyntaxToken()
            Dim isMustOverride = False
            Dim isNotOverridable = False
            Me._isSub = False
            Me._isFunction = False
            Me._isProperty = False

            Do While IsOnStartLine(token.SpanStart, text, startLine)
                Select Case token.Kind
                    Case SyntaxKind.OverridesKeyword
                        overridesToken = token
                    Case SyntaxKind.MustOverrideKeyword
                        isMustOverride = True
                    Case SyntaxKind.NotOverridableKeyword
                        isNotOverridable = True
                    Case SyntaxKind.FunctionKeyword
                        _isFunction = True
                    Case SyntaxKind.PropertyKeyword
                        _isProperty = True
                    Case SyntaxKind.SubKeyword
                        _isSub = True

                    ' Filter on accessibility by keeping the first one that we see
                    Case SyntaxKind.PublicKeyword
                        If seenAccessibility = Accessibility.NotApplicable Then
                            seenAccessibility = Accessibility.Public
                        End If

                    Case SyntaxKind.FriendKeyword
                        If seenAccessibility = Accessibility.NotApplicable Then
                            seenAccessibility = Accessibility.Internal
                        End If

                        ' If we see Friend AND Protected, assume Friend Protected
                        If seenAccessibility = Accessibility.Protected Then
                            seenAccessibility = Accessibility.ProtectedOrInternal
                        End If

                    Case SyntaxKind.ProtectedKeyword
                        If seenAccessibility = Accessibility.NotApplicable Then
                            seenAccessibility = Accessibility.Protected
                        End If

                        ' If we see Protected and Friend, assume Protected Friend
                        If seenAccessibility = Accessibility.Internal Then
                            seenAccessibility = Accessibility.ProtectedOrInternal
                        End If

                    Case Else
                        ' If we see anything else, give up
                        Return False
                End Select

                Dim previousToken = token.GetPreviousToken()

                ' Consume only modifiers on the same line
                If previousToken.Kind = SyntaxKind.None OrElse Not IsOnStartLine(previousToken.SpanStart, text, startLine) Then
                    Exit Do
                End If

                token = previousToken
            Loop

            modifiers = New DeclarationModifiers(isAbstract:=isMustOverride, isOverride:=True, isSealed:=isNotOverridable)
            Return overridesToken.Kind = SyntaxKind.OverridesKeyword AndAlso IsOnStartLine(overridesToken.Parent.SpanStart, text, startLine)
        End Function

        Public Overrides Function TryDetermineReturnType(startToken As SyntaxToken,
                                                         semanticModel As SemanticModel,
                                                         cancellationToken As CancellationToken,
                                                         ByRef returnType As ITypeSymbol, ByRef nextToken As SyntaxToken) As Boolean
            nextToken = startToken
            returnType = Nothing

            Return True
        End Function

        Public Overrides Function FilterOverrides(members As ImmutableArray(Of ISymbol),
                                                  returnType As ITypeSymbol) As ImmutableArray(Of ISymbol)
            ' Start by removing Finalize(), which we never want to show.
            Dim finalizeMethod = members.OfType(Of IMethodSymbol)().Where(Function(x) x.Name = "Finalize" AndAlso OverridesObjectMethod(x)).SingleOrDefault()
            If finalizeMethod IsNot Nothing Then
                members = members.Remove(finalizeMethod)
            End If

            If Me._isFunction Then
                ' Function: look for non-void return types
                Dim filteredMembers = members.OfType(Of IMethodSymbol)().Where(Function(m) Not m.ReturnsVoid)
                If filteredMembers.Any Then
                    Return ImmutableArray(Of ISymbol).CastUp(filteredMembers.ToImmutableArray())
                End If
            ElseIf Me._isProperty Then
                ' Property: return properties
                Dim filteredMembers = members.Where(Function(m) m.Kind = SymbolKind.Property)
                If filteredMembers.Any Then
                    Return filteredMembers.ToImmutableArray()
                End If
            ElseIf Me._isSub Then
                ' Sub: look for void return types
                Dim filteredMembers = members.OfType(Of IMethodSymbol)().Where(Function(m) m.ReturnsVoid)
                If filteredMembers.Any Then
                    Return ImmutableArray(Of ISymbol).CastUp(filteredMembers.ToImmutableArray())
                End If
            End If

            Return members.WhereAsArray(Function(m) Not m.IsKind(SymbolKind.Event))
        End Function

        Private Shared Function OverridesObjectMethod(method As IMethodSymbol) As Boolean
            Dim overriddenMember = method
            Do While overriddenMember.OverriddenMethod IsNot Nothing
                overriddenMember = overriddenMember.OverriddenMethod
            Loop

            If overriddenMember.ContainingType.SpecialType = SpecialType.System_Object Then
                Return True
            End If

            Return False
        End Function

        Protected Overrides Function GetTargetSelectionSpan(caretTarget As SyntaxNode) As TextSpan
            Dim node = DirectCast(caretTarget, SyntaxNode)

            ' MustOverride Sub | MustOverride Function: move to end of line
            Dim methodStatement = TryCast(node, MethodStatementSyntax)
            If methodStatement IsNot Nothing Then
                Return New TextSpan(methodStatement.GetLocation().SourceSpan.End, 0)
            End If

            Dim methodBlock = TryCast(node, MethodBlockBaseSyntax)
            If methodBlock IsNot Nothing Then
                Dim lastStatement = methodBlock.Statements.LastOrDefault()
                If lastStatement IsNot Nothing Then
                    Return New TextSpan(lastStatement.GetLocation().SourceSpan.End, 0)
                End If
            End If

            Dim propertyBlock = TryCast(node, PropertyBlockSyntax)
            If propertyBlock IsNot Nothing Then
                Dim firstAccessor = propertyBlock.Accessors.FirstOrDefault()
                If firstAccessor IsNot Nothing Then
                    Dim lastAccessorStatement = firstAccessor.Statements.LastOrDefault()
                    If lastAccessorStatement IsNot Nothing Then
                        Return New TextSpan(lastAccessorStatement.GetLocation().SourceSpan.End, 0)
                    End If
                End If
            End If

            Return New TextSpan(0, 0)
        End Function
    End Class
End Namespace

File: src/Features/CSharp/Portable/Completion/CompletionProviders/OverrideCompletionProvider.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Collections.Immutable;
using System.Composition;
using System.Diagnostics.CodeAnalysis;
using System.Threading;
using Microsoft.CodeAnalysis.Completion;
using Microsoft.CodeAnalysis.Completion.Providers;
using Microsoft.CodeAnalysis.CSharp.Extensions;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Editing;
using Microsoft.CodeAnalysis.Shared.Extensions;
using Microsoft.CodeAnalysis.Shared.Utilities;
using Microsoft.CodeAnalysis.Text;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers;

[ExportCompletionProvider(nameof(OverrideCompletionProvider), LanguageNames.CSharp)]
[ExtensionOrder(After = nameof(PreprocessorCompletionProvider))]
[Shared]
internal partial class OverrideCompletionProvider : AbstractOverrideCompletionProvider
{
    [ImportingConstructor]
    [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
    public OverrideCompletionProvider()
    {
    }

    internal override string Language => LanguageNames.CSharp;

    protected override SyntaxNode GetSyntax(SyntaxToken token)
    {
        return token.GetAncestor<EventFieldDeclarationSyntax>()
            ?? token.GetAncestor<EventDeclarationSyntax>()
            ?? token.GetAncestor<PropertyDeclarationSyntax>()
            ?? token.GetAncestor<IndexerDeclarationSyntax>()
            ?? (SyntaxNode?)token.GetAncestor<MethodDeclarationSyntax>()
            ?? throw ExceptionUtilities.UnexpectedValue(token);
    }

    public override bool IsInsertionTrigger(SourceText text, int characterPosition, CompletionOptions options)
        => CompletionUtilities.IsTriggerAfterSpaceOrStartOfWordCharacter(text, characterPosition, options);

    public override ImmutableHashSet<char> TriggerCharacters { get; } = CompletionUtilities.SpaceTriggerCharacter;

    protected override SyntaxToken GetToken(CompletionItem completionItem, SyntaxTree tree, CancellationToken cancellationToken)
    {
        var tokenSpanEnd = MemberInsertionCompletionItem.GetTokenSpanEnd(completionItem);
        return tree.FindTokenOnLeftOfPosition(tokenSpanEnd, cancellationToken);
    }

    public override bool TryDetermineReturnType(SyntaxToken startToken, SemanticModel semanticModel, CancellationToken cancellationToken, out ITypeSymbol? returnType, out SyntaxToken nextToken)
    {
        nextToken = startToken;
        returnType = null;
        if (startToken.Parent is TypeSyntax typeSyntax)
        {
            // 'partial' is actually an identifier.  If we see it just bail.  This does mean
            // we won't handle overrides that actually return a type called 'partial'.  And
            // not a single tear was shed.
            if (typeSyntax is IdentifierNameSyntax identifierName &&
                identifierName.Identifier.IsKindOrHasMatchingText(SyntaxKind.PartialKeyword))
            {
                return false;
            }

            returnType = semanticModel.GetTypeInfo(typeSyntax, cancellationToken).Type;
            nextToken = typeSyntax.GetFirstToken().GetPreviousToken();
        }

        return true;
    }

    public override bool TryDetermineModifiers(
        SyntaxToken startToken,
        SourceText text,
        int startLine,
        out Accessibility seenAccessibility,
        out DeclarationModifiers modifiers)
    {
        var token = startToken;
        var parentMember = token.Parent;
        modifiers = default;
        seenAccessibility = Accessibility.NotApplicable;

        if (parentMember is null)
            return false;

        // Keep walking backwards as long as we're still within our parent member.
        while (token != default)
        {
            if (token.SpanStart < parentMember.SpanStart)
            {
                // moved before the start of the member we're in.  If previous member's token is on the same line,
                // we bail out as our replacement will delete the entire line we're on.
                if (IsOnStartLine(token.SpanStart, text, startLine))
                    return false;

                break;
            }

            // Ok to hit a `]` if it's the end of attributes on this member.
            if (token.Kind() == SyntaxKind.CloseBracketToken)
            {
                if (token.Parent is not AttributeListSyntax)
                    return false;

                break;
            }

            // We only accept tokens that precede us on the same line.  Splitting across multiple lines is too niche
            // to want to support, and more likely indicates a case of broken code that the user is in the middle of
            // fixing up.
            if (!IsOnStartLine(token.SpanStart, text, startLine))
                return false;

            switch (token.Kind())
            {
                // Standard modifier cases we accept.

                case SyntaxKind.AbstractKeyword:
                    modifiers = modifiers.WithIsAbstract(true);
                    break;
                case SyntaxKind.ExternKeyword:
                    modifiers = modifiers.WithIsExtern(true);
                    break;
                case SyntaxKind.OverrideKeyword:
                    modifiers = modifiers.WithIsOverride(true);
                    break;
                case SyntaxKind.RequiredKeyword:
                    modifiers = modifiers.WithIsRequired(true);
                    break;
                case SyntaxKind.SealedKeyword:
                    modifiers = modifiers.WithIsSealed(true);
                    break;
                case SyntaxKind.UnsafeKeyword:
                    modifiers = modifiers.WithIsUnsafe(true);
                    break;

                // Accessibility modifiers we accept.

                case SyntaxKind.PublicKeyword:
                    seenAccessibility = seenAccessibility == Accessibility.NotApplicable
                        ? Accessibility.Public
                        : seenAccessibility;
                    break;
                case SyntaxKind.PrivateKeyword:
                    seenAccessibility = seenAccessibility switch
                    {
                        Accessibility.NotApplicable => Accessibility.Private,
                        // If we see private AND protected, filter for private protected
                        Accessibility.Protected => Accessibility.ProtectedAndInternal,
                        _ => seenAccessibility,
                    };
                    break;
                case SyntaxKind.InternalKeyword:
                    // If we see internal AND protected, filter for protected internal
                    seenAccessibility = seenAccessibility switch
                    {
                        Accessibility.NotApplicable => Accessibility.Internal,
                        Accessibility.Protected => Accessibility.ProtectedOrInternal,
                        _ => seenAccessibility,
                    };
                    break;
                case SyntaxKind.ProtectedKeyword:
                    seenAccessibility = seenAccessibility switch
                    {
                        Accessibility.NotApplicable => Accessibility.Protected,
                        // If we see protected AND internal, filter for protected internal.
                        Accessibility.Internal => Accessibility.ProtectedOrInternal,
                        // Or if we see private AND protected, filter for private protected
                        Accessibility.Private => Accessibility.ProtectedAndInternal,
                        _ => seenAccessibility,
                    };
                    break;
                default:
                    // If we hit anything else, then this token is not valid for override completions, and we can just bail here.
                    return false;
            }

            token = token.GetPreviousToken();
        }

        // Have to at least found the override token for us to offer override-completion.
        return modifiers.IsOverride;
    }

    public override SyntaxToken FindStartingToken(SyntaxTree tree, int position, CancellationToken cancellationToken)
    {
        var token = tree.FindTokenOnLeftOfPosition(position, cancellationToken);
        return token.GetPreviousTokenIfTouchingWord(position);
    }

    public override ImmutableArray<ISymbol> FilterOverrides(ImmutableArray<ISymbol> members, ITypeSymbol? returnType)
    {
        if (returnType == null)
        {
            return members;
        }

        var filteredMembers = members.WhereAsArray(m =>
            SymbolEquivalenceComparer.Instance.Equals(GetReturnType(m), returnType));

        // Don't filter by return type if we would then have nothing to show.
        // This way, the user gets completion even if they speculatively typed the wrong return type
        return filteredMembers.Length > 0 ? filteredMembers : members;
    }

    protected override int GetTargetCaretPosition(SyntaxNode caretTarget)
    {
        // Inserted Event declarations are a single line, so move to the end of the line.
        if (caretTarget is EventFieldDeclarationSyntax)
        {
            return caretTarget.GetLocation().SourceSpan.End;
        }
        else if (caretTarget is MethodDeclarationSyntax methodDeclaration)
        {
            return CompletionUtilities.GetTargetCaretPositionForMethod(methodDeclaration);
        }
        else if (caretTarget is BasePropertyDeclarationSyntax propertyDeclaration)
        {
            // property: no accessors; move to the end of the declaration
            if (propertyDeclaration.AccessorList != null && propertyDeclaration.AccessorList.Accessors.Any())
            {
                // move to the end of the last statement of the first accessor
                var firstAccessor = propertyDeclaration.AccessorList.Accessors[0];
                var firstAccessorStatement = (SyntaxNode?)firstAccessor.Body?.Statements.LastOrDefault() ??
                    firstAccessor.ExpressionBody!.Expression;
                return firstAccessorStatement.GetLocation().SourceSpan.End;
            }
            else
            {
                return propertyDeclaration.GetLocation().SourceSpan.End;
            }
        }
        else
        {
            throw ExceptionUtilities.Unreachable();
        }
    }
}
